### ISSUE TITLE
Cloud friendly file storage

### ISSUE THREAD

**Issue by rjackson** [tfwiki/deployment#5]
We're currently using NFS, as that's what we use for the existing site, but it's not exactly a cloud-friendly technology, and it requires our config hard-code a service IP address when running in Google Container Engine (kubernetes/kubernetes#48212)

We should look into Cloud-friendly alternatives, particularly those that are easy and quick to migrate to (i.e. minimal disruption to production).

----

_September 2018: Thinking-through-writing about Object Storage_

NFS as a ReadWriteMany volume has served fairly well since the Google Cloud migration, but its not been entirely reliable: We get occasional file-system related fatal errors (NFS erroneously reporting directories as not writable springs to mind), and there have been a couple instances of old image cache entries hanging around after overwrites. On top of that, I'm not confident in my ability to maintain NFS should any issues arise; it works now, but if that were to change I wouldn't know where to start digging.

I would rather we were persisting images into object storage, as is typical of modern cloud-deployed systems these days. As we're running on GKE, Google Cloud Storage ("GCS") is the obvious candidate.

Unfortunately, MediaWiki doesn't speak Object Storage natively, beyond [`SwiftFileBackend`](https://doc.wikimedia.org/mediawiki-core/master/php/classSwiftFileBackend.html). It speaks Filesystem, or there are years-old extensions implementing `FileRepo` for [Amazon S3](https://www.mediawiki.org/wiki/Extension:LocalS3Repo) and [Windows Azure Storage](https://www.mediawiki.org/wiki/Extension:WindowsAzureStorage).

Another unfortunate scenario is the fact Kubernetes' volumes doesn't speak Object Storage either (or maybe it does through some of its storage drivers, but it doesn't speak GCS). There is an open issue for FUSE volumes (kubernetes/kubernetes#7890), which would help: We can mount GCS via FUSE. Until then, there are some workarounds mentioned in that issue; the most promising being a DaemonSet to mount the FUSE volume on the host, and then mounting the host path on the container.

So our options for Cloud Friendly (object) storage are:

* Write a Google Cloud Storage extension for MediaWiki
* Host-mounted FUSE volume via DaemonSet

I'm leaning more toward the first option at the moment, as the S3 and Azure extensions are available to serve as reference (assuming those extensions haven't been touched in years because they "just work", and not that they've been abandoned 😬).

The FUSE-in-Kubernetes workaround feels like we'd just be replacing NFS layers with FUSE layers:

```diff
-MediaWiki -> FSFileBackend -> `/uploads` -> (NFS volume) -> disk
+MediaWiki -> FSFileBackend -> `/uploads` -> (FUSE workaround) -> GCS
```

Which would put me back in the same situation: Extra complexity I'm not familiar handling.

Let's set our sights on a new extension, so the flow is more like:

```diff
-MediaWiki -> FSFileBackend -> `/uploads` -> NFS -> disk
+MediaWiki -> GCSFileBackend -> GCS
```

**Comment by jingxu97** [comment]
You can use service name now, but it has to be the full DNS name.

**Comment by rjackson** [comment]
@jingxu97 Ooh perfect, thanks for letting me know!

**Comment by rjackson** [comment]
I've expanded the original issue's description with thoughts on replacing NFS with Google Cloud Storage.

**Comment by rjackson** [comment]
Currently exploring Google Cloud Filestore – a managed NFS service. It's pricey, but I assume will be better looked after than leaving an old "nfs-server" Docker image floating around untouched for a couple of years 🙃

**Comment by rjackson** [comment]
Implemented Filestore e6cee8ca036c11f849ffb52876ebf155fdf25620...ffc7ef5bb69d8b3706d5d096295a1300c59f64f0
